app/scripts/controllers/plugins.js

```diff
@@ -49,43 +64,79 @@ class PluginsController extends EventEmitter {
     this._blockTracker = opts._blockTracker
     this._getAccounts = opts._getAccounts
     this._removeAllPermissionsFor = opts._removeAllPermissionsFor
+    this._getPermissionsFor = opts._getPermissionsFor
     this.getApi = opts.getApi
     this.getAppKeyForDomain = opts.getAppKeyForDomain
     this.onUnlock = opts.onUnlock
+    this.closeAllConnections = opts.closeAllConnections
 
     this.rpcMessageHandlers = new Map()
     this.apiRequestHandlers = new Map()
     this.adding = {}
   }
 
   runExistingPlugins () {
+
     const plugins = this.store.getState().plugins
+
     if (Object.keys(plugins).length > 0) {
       console.log('running existing plugins', plugins)
     } else {
       console.log('no plugins found on boot')
       return
     }
 
-    Object.values(plugins).forEach(({ pluginName, approvedPermissions, sourceCode }) => {
+    Object.values(plugins).forEach(({ name: pluginName, sourceCode }) => {
+
       console.log(`running: ${pluginName}`)
-      const ethereumProvider = this.setupProvider({ hostname: pluginName }, async () => {
-        return { name: pluginName }
-      }, true)
+      const approvedPermissions = this._getPermissionsFor(pluginName).map(
+        perm => perm.parentCapability
+      )
+
+      const ethereumProvider = this.setupProvider(
+        { hostname: pluginName },
+        getGetDomainMetadataFunction(pluginName),
```

Comment:
Maybe I'd name it `createGetDomainMetadataFunction` or something, since this function doesn't exist yet, so it's not quite a getter? Avoids the redundancy, anyways..

Reply:
Yeah, that's much better!